external/avro/src/main/scala/org/apache/spark/sql/avro/AvroDataToCatalyst.scala

```diff
@@ -53,7 +54,9 @@ case class AvroDataToCatalyst(
 
   @transient private lazy val avroSchema = new Schema.Parser().parse(jsonFormatSchema)
 
-  @transient private lazy val reader = new GenericDatumReader[Any](avroSchema)
```

Comment:
do you mean the reader must use a schema that matches the avro encoded binary, but we can use a compatible schema in `AvroDeserializer`, which can do schema evolution?

Reply:
Exactly, to read data you can use a compatible evolved schema, but you also need to pass the schema with which the data was originally serialized. Using this GenericDatumReader API https://avro.apache.org/docs/1.8.2/api/java/org/apache/avro/generic/GenericDatumReader.html#GenericDatumReader(org.apache.avro.Schema,%20org.apache.avro.Schema)